File: apps/web/src/lib/debris.ts
// Server-only: load tracked debris (live engine, else a direct CelesTrak pull,
// else the bundled snapshot) and compute per-object collision risk.
//
// Mirrors lib/catalog.ts's resilient-fallback philosophy so the debris layer
// always has data: the live engine /debris is preferred (it has DISCOS-enriched
// RCS + the screening loop's real Pc), then a direct server-side fetch of the
// CelesTrak debris fragmentation groups (server→server, so the browser's strict
// `connect-src 'self'` CSP doesn't apply), then a committed snapshot for
// offline/CI. Imports the snapshot JSON, so this file must only be pulled into
// server route handlers — never a client component.

import debrisData from "@/data/debris.json";
import {
  fetchDebris,
  type Debris,
  type DebrisRiskSummary,
} from "@/lib/orbital-engine";
import { deriveApsis } from "@/app/utils/sgp4FromTle";
import {
  circularSpeedKmS,
  riskLevelForScore,
  riskScore,
} from "@/app/data/debrisRisk";

const CELESTRAK_GP_URL = "https://celestrak.org/NORAD/elements/gp.php";

// The three major fragmentation clouds (~1,150 tracked objects together; ≫200).
// Matches the engine's `celestrak_debris_groups` default.
const DEBRIS_GROUPS = [
  "fengyun-1c-debris",
  "cosmos-2251-debris",
  "iridium-33-debris",
];

// Below this the engine/CelesTrak result is treated as degenerate (unseeded DB,
// blocked network) and we fall through to the next source.
const MIN_LIVE_DEBRIS = 20;

/** Raw debris record — the bundled-snapshot shape + what the TLE parse yields. */
interface DebrisRecord {
  OBJECT_NAME: string;
  OBJECT_ID?: string | null;
  NORAD_CAT_ID?: number | null;
  COUNTRY_CODE?: string | null;
  RCS_SIZE?: string | null;
  TLE_LINE0?: string | null;
  TLE_LINE1: string;
  TLE_LINE2: string;
}

/** USOID for a debris object, matching the engine's make_usoid (CAT precedence). */
function usoidFor(norad: number | null, name: string): string {
  return norad != null ? `SH:CAT:${String(norad).padStart(9, "0")}` : name;
}

/** Compute orbital params + risk for a raw debris record (TLE-only path). */
function toDebris(rec: DebrisRecord): Debris | null {
  if (!rec.TLE_LINE1 || !rec.TLE_LINE2) return null;
  const apsis = deriveApsis(rec.TLE_LINE1, rec.TLE_LINE2);
  if (apsis === null) return null;
  const meanAlt = apsis.meanAltitudeKm;
  const speed = circularSpeedKmS(meanAlt);
  const score = riskScore(meanAlt, speed, rec.RCS_SIZE);
  const norad =
    rec.NORAD_CAT_ID ?? (Number.parseInt(rec.TLE_LINE1.slice(2, 7), 10) || null);
  return {
    object_id: rec.OBJECT_ID ?? usoidFor(norad, rec.OBJECT_NAME),
    norad_cat_id: norad,
    object_name: rec.OBJECT_NAME,
    country_code: rec.COUNTRY_CODE ?? null,
    rcs_size: rec.RCS_SIZE ?? null,
    inclination: apsis.inclinationDeg,
    eccentricity: apsis.eccentricity,
    period_min: apsis.periodMin,
    apoapsis_km: apsis.apogeeKm,
    periapsis_km: apsis.perigeeKm,
    mean_altitude_km: meanAlt,
    speed_km_s: speed,
    risk_score: score,
    risk_level: riskLevelForScore(score),
    tle_line0: rec.TLE_LINE0 ?? rec.OBJECT_NAME,
    tle_line1: rec.TLE_LINE1,
    tle_line2: rec.TLE_LINE2,
  };
}

/** Parse a CelesTrak 3-line TLE export into raw records (one per 3 lines). */
function parseTleText(text: string): DebrisRecord[] {
  const lines = text
    .split(/\r?\n/)
    .map((l) => l.trimEnd())
    .filter((l) => l.length > 0);
  const out: DebrisRecord[] = [];
  for (let i = 0; i + 2 < lines.length; i += 3) {
    const l0 = lines[i];
    const l1 = lines[i + 1];
    const l2 = lines[i + 2];
    if (!l1?.startsWith("1 ") || !l2?.startsWith("2 ")) continue;
    out.push({
      OBJECT_NAME: l0.replace(/^0 /, "").trim(),
      NORAD_CAT_ID: Number.parseInt(l1.slice(2, 7), 10) || null,
      TLE_LINE0: l0,
      TLE_LINE1: l1,
      TLE_LINE2: l2,
    });
  }
  return out;
}

/** Direct server-side fetch of the CelesTrak debris groups (one TLE pull each). */
async function fetchCelestrakDebris(): Promise<Debris[]> {
  const out: Debris[] = [];
  for (const group of DEBRIS_GROUPS) {
    try {
      const res = await fetch(
        `${CELESTRAK_GP_URL}?GROUP=${group}&FORMAT=tle`,
        { cache: "no-store" }
      );
      if (!res.ok) continue; // skip a throttled/failed group, keep the rest
      for (const rec of parseTleText(await res.text())) {
        const d = toDebris(rec);
        if (d !== null) out.push(d);
      }
    } catch {
      /* skip this group */
    }
  }
  return out;
}

// Committed snapshot of the debris groups (built by scripts/build-debris-snapshot.mjs)
// — the offline/CI fallback. Risk is computed here so the snapshot stays compact.
const fallbackDebris: Debris[] = (debrisData as DebrisRecord[])
  .map(toDebris)
  .filter((d): d is Debris => d !== null);

/**
 * The tracked-debris population with risk: the live engine when healthy, else a
 * direct CelesTrak pull, else the bundled snapshot. Never throws — the layer
 * always gets data.
 */
export async function loadDebris(): Promise<Debris[]> {
  try {
    const live = await fetchDebris();
    if (live !== null && live.length >= MIN_LIVE_DEBRIS) return live;
  } catch {
    /* fall through */
  }
  try {
    const celestrak = await fetchCelestrakDebris();
    if (celestrak.length >= MIN_LIVE_DEBRIS) return celestrak;
  } catch {
    /* fall through */
  }
  return fallbackDebris;
}

const RISK_LEVELS = ["Critical", "High", "Medium", "Low"];

/**
 * Population risk analysis computed in-process from a debris set — the offline
 * fallback for `/api/debris/risk` when the engine (which also returns screened
 * conjunctions) is unavailable. Counts per risk level + a debris-density-by-100km
 * -shell profile; `conjunctions` stays empty (real Pc needs the screening engine).
 */
export function summarizeDebrisRisk(debris: Debris[]): DebrisRiskSummary {
  const byLevel: Record<string, number> = Object.fromEntries(
    RISK_LEVELS.map((l) => [l, 0])
  );
  const shells = new Map<number, number>();
  for (const d of debris) {
    byLevel[d.risk_level] = (byLevel[d.risk_level] ?? 0) + 1;
    if (d.mean_altitude_km !== null) {
      const band = Math.floor(d.mean_altitude_km / 100) * 100;
      shells.set(band, (shells.get(band) ?? 0) + 1);
    }
  }
  const density = [...shells.entries()]
    .sort((a, b) => a[0] - b[0])
    .map(([altitude_km, count]) => ({ altitude_km, count }));
  return {
    available: true,
    total: debris.length,
    by_level: byLevel,
    density_by_shell: density,
    conjunctions: [],
  };
}

File: services/orbital-engine/orbital_engine/api/debris.py
"""Debris catalog + risk endpoints consumed by the web BFF (debris-viz/risk).

``GET /debris`` returns the tracked-debris subset of the catalog with per-object
risk (the globe layer + popups). ``GET /debris/risk`` adds the population analysis:
counts per risk level, a debris-density-by-altitude-shell profile (the classic
ODPO LEO profile, peaking ~800-1000 km), and the screened debris conjunctions
with their real probability-of-collision (reusing the screening loop's output).

Risk *level* comes from the explainable ``debris_risk`` model (always available);
collision *probability* comes from the engine's SGP4 screening (``estimate_pc``)
surfaced via the conjunctions — the two are deliberately distinct.
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, Field

from orbital_engine.api.conjunctions import ConjunctionResponse
from orbital_engine.debris_risk import circular_speed_km_s, risk_level, risk_score
from orbital_engine.repository import query_debris, query_debris_conjunctions
from orbital_engine.screening import orbit_altitudes
from orbital_engine.security.attributes import Action, DataDomain, Subject
from orbital_engine.security.enforce import requires

router = APIRouter(tags=["debris"])


class DebrisObject(BaseModel):
    """One tracked-debris object with derived orbital params + risk."""

    object_id: str
    norad_cat_id: int | None = None
    object_name: str
    country_code: str | None = None
    rcs_size: str | None = None
    inclination: float | None = None
    eccentricity: float | None = None
    period_min: float | None = None
    apoapsis_km: float | None = None
    periapsis_km: float | None = None
    mean_altitude_km: float | None = None
    speed_km_s: float | None = None
    risk_score: float = 0.0
    risk_level: str = "Low"
    tle_line0: str | None = None
    tle_line1: str | None = None
    tle_line2: str | None = None


class ShellBin(BaseModel):
    """Debris count in one 100 km altitude shell (density-by-shell profile)."""

    altitude_km: int
    count: int


class DebrisRiskSummary(BaseModel):
    """Population-level debris risk analysis."""

    total: int
    by_level: dict[str, int] = Field(default_factory=dict)
    density_by_shell: list[ShellBin] = Field(default_factory=list)
    conjunctions: list[ConjunctionResponse] = Field(default_factory=list)


def _enrich(row: dict[str, Any]) -> dict[str, Any]:
    """Derive mean altitude, speed and the risk score/level for one debris row."""
    apo = row.get("apoapsis_km")
    peri = row.get("periapsis_km")
    mm = row.get("mean_motion")
    ecc = row.get("eccentricity")
    if (apo is None or peri is None) and mm is not None and ecc is not None:
        apo, peri = orbit_altitudes(float(mm), float(ecc))
    mean_alt = (float(apo) + float(peri)) / 2.0 if apo is not None and peri is not None else None
    speed = circular_speed_km_s(mean_alt) if mean_alt is not None else None
    score = (
        risk_score(mean_alt, speed, row.get("rcs_size"))
        if mean_alt is not None and speed is not None
        else 0.0
    )
    return {
        **row,
        "apoapsis_km": apo,
        "periapsis_km": peri,
        "mean_altitude_km": mean_alt,
        "speed_km_s": speed,
        "risk_score": score,
        "risk_level": risk_level(score),
    }


@router.get("/debris", response_model=list[DebrisObject], summary="Tracked debris + risk")
async def get_debris(
    limit: int = Query(default=2000, ge=1, le=20000),
    offset: int = Query(default=0, ge=0),
    _subject: Subject = Depends(requires(DataDomain.CATALOG, Action.QUERY)),
) -> list[dict[str, Any]]:
    return [_enrich(row) for row in await query_debris(limit=limit, offset=offset)]


@router.get("/debris/risk", response_model=DebrisRiskSummary, summary="Debris risk analysis")
async def get_debris_risk(
    _subject: Subject = Depends(requires(DataDomain.SCREENING, Action.QUERY)),
) -> DebrisRiskSummary:
    rows = [_enrich(row) for row in await query_debris(limit=20000)]
    by_level: dict[str, int] = {"Critical": 0, "High": 0, "Medium": 0, "Low": 0}
    shells: dict[int, int] = {}
    for row in rows:
        by_level[row["risk_level"]] = by_level.get(row["risk_level"], 0) + 1
        alt = row.get("mean_altitude_km")
        if alt is not None:
            band = int(alt // 100) * 100
            shells[band] = shells.get(band, 0) + 1
    density = [ShellBin(altitude_km=k, count=shells[k]) for k in sorted(shells)]
    conjunctions = await query_debris_conjunctions(limit=200)
    return DebrisRiskSummary(
        total=len(rows),
        by_level=by_level,
        density_by_shell=density,
        conjunctions=conjunctions,  # type: ignore[arg-type]
    )
